File: mobile/ios/FUN/FUN/Models/Series.swift
/**
 * FUN App - Series Model
 */

import Foundation

struct Creator: Codable, Identifiable {
    let id: String
    let displayName: String
    let profileImage: String
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case displayName, profileImage
    }
}

struct Season: Codable, Identifiable {
    let id: String
    let seasonNumber: Int
    let title: String?
    let description: String?
    let episodes: [Episode]
    let thumbnailUrl: String?
    let releaseDate: Date?
    var isCompleted: Bool? // User has watched all episodes
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case seasonNumber, title, description, episodes
        case thumbnailUrl, releaseDate, isCompleted
    }
}

struct Series: Codable, Identifiable {
    let id: String
    let title: String
    let description: String
    let thumbnailUrl: String
    let coverImageUrl: String
    let genre: [String]
    let tags: [String]
    let creatorId: String
    let creator: Creator
    let totalEpisodes: Int
    let totalViews: Int
    let totalLikes: Int
    let totalComments: Int
    let isActive: Bool
    let isFeatured: Bool
    let createdAt: Date
    let seasons: [Season]? // Multi-season support
    let episodes: [Episode]? // Backward compatibility
    
    // Legacy support
    let rating: Double?
    let isLiked: Bool?
    let isFavorited: Bool?
    
    var stats: SeriesStats {
        SeriesStats(
            totalViews: totalViews,
            totalLikes: totalLikes,
            totalComments: totalComments
        )
    }
    
    var allEpisodes: [Episode] {
        if let seasons = seasons {
            return seasons
                .sorted { $0.seasonNumber < $1.seasonNumber }
                .flatMap { $0.episodes.sorted { $0.episodeNumber < $1.episodeNumber } }
        } else if let episodes = episodes {
            return episodes.sorted { $0.episodeNumber < $1.episodeNumber }
        }
        return []
    }
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case title, description, thumbnailUrl, coverImageUrl
        case genre, tags, creatorId, creator
        case totalEpisodes, totalViews, totalLikes, totalComments
        case isActive, isFeatured, createdAt, seasons, episodes
        case rating, isLiked, isFavorited
    }
}

struct SeriesStats: Codable {
    let totalViews: Int
    let totalLikes: Int
    let totalComments: Int
}

struct SeriesListResponse: Codable {
    let series: [Series]
    let pagination: Pagination
}

struct SeriesDetailResponse: Codable {
    let series: Series
}

struct Pagination: Codable {
    let page: Int
    let limit: Int
    let total: Int
    let pages: Int
}

File: mobile/android/FUN/app/src/main/java/com/fun/app/data/models/Series.kt
/**
 * FUN App - Series Models
 */

package com.fun.app.data.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class Creator(
    @Json(name = "_id") val id: String,
    val displayName: String,
    val profileImage: String
)

@JsonClass(generateAdapter = true)
data class Season(
    @Json(name = "_id") val id: String,
    val seasonNumber: Int,
    val title: String?,
    val description: String?,
    val episodes: List<Episode>,
    val thumbnailUrl: String?,
    val releaseDate: java.util.Date?,
    var isCompleted: Boolean? = null // User has watched all episodes
)

@JsonClass(generateAdapter = true)
data class Series(
    @Json(name = "_id") val id: String,
    val title: String,
    val description: String,
    val thumbnailUrl: String,
    val coverImageUrl: String,
    val genre: List<String>,
    val tags: List<String>,
    val creatorId: String,
    val creator: Creator,
    val totalEpisodes: Int,
    val stats: SeriesStats,
    val isActive: Boolean,
    val isFeatured: Boolean,
    val createdAt: java.util.Date,
    val seasons: List<Season>? = null, // Multi-season support
    val episodes: List<Episode>? = null, // Backward compatibility
    // Legacy support
    val rating: Double? = null,
    val isLiked: Boolean? = null,
    val isFavorited: Boolean? = null
) {
    val allEpisodes: List<Episode>
        get() {
            return seasons
                ?.sortedBy { it.seasonNumber }
                ?.flatMap { it.episodes.sortedBy { ep -> ep.episodeNumber } }
                ?: episodes?.sortedBy { it.episodeNumber }
                ?: emptyList()
        }
}

@JsonClass(generateAdapter = true)
data class SeriesStats(
    val totalViews: Int,
    val totalLikes: Int,
    val totalComments: Int
)

@JsonClass(generateAdapter = true)
data class SeriesListResponse(
    val series: List<Series>,
    val pagination: Pagination
)

@JsonClass(generateAdapter = true)
data class SeriesDetailResponse(
    val series: Series
)

@JsonClass(generateAdapter = true)
data class Pagination(
    val page: Int,
    val limit: Int,
    val total: Int,
    val pages: Int
)
